src/main/java/com/conveyal/r5/profile/FastRaptorWorker.java

```diff
@@ -0,0 +1,559 @@
+    private void prefilterPatterns () {
+        TIntList frequencyPatterns = new TIntArrayList();
+        TIntList scheduledPatterns = new TIntArrayList();
+        frequencyIndexForOriginalPatternIndex = new int[transit.tripPatterns.size()];
+        Arrays.fill(frequencyIndexForOriginalPatternIndex, -1);
+        scheduledIndexForOriginalPatternIndex = new int[transit.tripPatterns.size()];
+        Arrays.fill(scheduledIndexForOriginalPatternIndex, -1);
+
+        int patternIndex = -1; // first increment lands at 0
+        int frequencyIndex = 0;
+        int scheduledIndex = 0;
+        for (TripPattern pattern : transit.tripPatterns) {
+            patternIndex++;
+            if (pattern.servicesActive.intersects(servicesActive)) {
+                // at least one trip on this pattern is relevant
+                if (pattern.hasFrequencies) {
+                    frequencyPatterns.add(patternIndex);
+                    frequencyIndexForOriginalPatternIndex[patternIndex] = frequencyIndex++;
+                }
+                if (pattern.hasSchedules) { // NB not else b/c we still support combined frequency and schedule patterns.
+                    scheduledPatterns.add(patternIndex);
+                    scheduledIndexForOriginalPatternIndex[patternIndex] = scheduledIndex++;
+                }
+            }
+        }
+
+        originalPatternIndexForFrequencyIndex = frequencyPatterns.toArray();
+        originalPatternIndexForScheduledIndex = scheduledPatterns.toArray();
+
+        runningFrequencyPatterns = IntStream.of(originalPatternIndexForFrequencyIndex)
+                .mapToObj(transit.tripPatterns::get).toArray(TripPattern[]::new);
+        runningScheduledPatterns = IntStream.of(originalPatternIndexForScheduledIndex)
+                .mapToObj(transit.tripPatterns::get).toArray(TripPattern[]::new);
+
+        LOG.info("Prefiltering patterns based on date active reduced {} patterns to {} frequency and {} scheduled patterns",
+                transit.tripPatterns.size(), frequencyPatterns.size(), scheduledPatterns.size());
+    }
+
+    /** Set the departure time in the scheduled search to the given departure time */
+    private void updateDepartureTime (int departureTime) {
+        for (RaptorState state : this.scheduleState) {
+            state.setDepartureTime(departureTime);
+
+            // clear all touched stops to avoid constant reëxploration
+            state.bestStopsTouched.clear();
+            state.nonTransferStopsTouched.clear();
+            // TODO prune trips that are now longer than max lengths to avoid biasing averages
+        }
+
+
+        // add initial stops
+        RaptorState initialState = scheduleState[0];
+        accessStops.forEachEntry((stop, accessTime) -> {
+            initialState.setTimeAtStop(stop, accessTime + departureTime, -1, -1, 0, 0, true);
+            return true; // continue iteration
+        });
+    }
+
+    /** Perform one minute of a RAPTOR search */
```

Comment:
Also please explain what is in the int[][] returned by this function. Looks like it would be an array of n arrays of travel times to every stop, where n= iterationsPerMinute.